pkg/virt-controller/watch/pool.go

```diff
@@ -0,0 +1,980 @@
+				errChan <- err
+				return
+			}
+			c.recorder.Eventf(pool, k8score.EventTypeNormal, SuccessfulDeleteVirtualMachineReason, "Deleted VM %s/%s with uid %v from pool", vm.Namespace, vm.Name, vm.ObjectMeta.UID)
+			log.Log.Object(pool).Infof("Deleted vm %s/%s from pool", vm.Namespace, vm.Name)
+		}(i)
+	}
+
+	wg.Wait()
+
+	select {
+	case err := <-errChan:
+		// Only return the first error which occurred. We log the rest
+		return err
+	default:
+	}
+
+	return nil
+}
+
+func generateVMName(index int, baseName string) string {
+	return fmt.Sprintf("%s-%d", baseName, index)
+}
+
+func calculateNewVMNames(count int, baseName string, namespace string, vmStore cache.Store) []string {
+	var newNames []string
+
+	// generate `count` new unused VM names
+	curIndex := 0
+	for n := 0; n < count; n++ {
+		// find next unused index starting where we left off last
+		i := curIndex
+		for {
+			name := generateVMName(i, baseName)
+			vmKey := controller.NamespacedKey(namespace, name)
+			_, exists, _ := vmStore.GetByKey(vmKey)
+			if !exists {
+				newNames = append(newNames, name)
+				curIndex = i + 1
+				break
+			}
+
+			i++
+		}
+	}
+
+	return newNames
+}
+
+func poolOwnerRef(pool *poolv1.VirtualMachinePool) metav1.OwnerReference {
+	t := pointer.BoolPtr(true)
+	gvk := schema.GroupVersionKind{Group: poolv1.SchemeGroupVersion.Group, Version: poolv1.SchemeGroupVersion.Version, Kind: poolv1.VirtualMachinePoolKind}
+	return metav1.OwnerReference{
+		APIVersion:         gvk.GroupVersion().String(),
+		Kind:               gvk.Kind,
+		Name:               pool.ObjectMeta.Name,
+		UID:                pool.ObjectMeta.UID,
+		Controller:         t,
+		BlockOwnerDeletion: t,
+	}
+}
+
+func indexFromName(name string) (int, error) {
+	slice := strings.Split(name, "-")
+	return strconv.Atoi(slice[len(slice)-1])
+}
+
+func indexVMSpec(spec *virtv1.VirtualMachineSpec, idx int) *virtv1.VirtualMachineSpec {
+
+	if len(spec.DataVolumeTemplates) == 0 {
+		return spec
+	}
+
+	dvNameMap := map[string]string{}
+	for i, _ := range spec.DataVolumeTemplates {
+
+		indexName := fmt.Sprintf("%s-%d", spec.DataVolumeTemplates[i].Name, idx)
+		dvNameMap[spec.DataVolumeTemplates[i].Name] = indexName
+
+		spec.DataVolumeTemplates[i].Name = indexName
+	}
+
+	for i, volume := range spec.Template.Spec.Volumes {
+		if volume.VolumeSource.PersistentVolumeClaim != nil {
+			indexName, ok := dvNameMap[volume.VolumeSource.PersistentVolumeClaim.ClaimName]
+			if ok {
+				spec.Template.Spec.Volumes[i].PersistentVolumeClaim.ClaimName = indexName
+			}
+		} else if volume.VolumeSource.DataVolume != nil {
+			indexName, ok := dvNameMap[volume.VolumeSource.DataVolume.Name]
+			if ok {
+				spec.Template.Spec.Volumes[i].DataVolume.Name = indexName
+			}
+		}
+	}
+
+	return spec
+}
+
+func injectHashIntoVM(vm *virtv1.VirtualMachine, vmHash string, vmiHash string) *virtv1.VirtualMachine {
```

Comment:
From what I could tell, you're using the hash as a checksum to monitor object change.   What was your thinking behind using this method over using something like deep equal?  Also, since this is a label, what will happen if a user modifies the label?

Reply:
>From what I could tell, you're using the hash as a checksum to monitor object change. What was your thinking behind using this method over using something like deep equal? 

I took this idea from the kubernetes controller deployment.  They do pretty much the same thing there. That's why you'll see labels like ```pod-template-hash: 96b6c9d98``` on pods created by a deployment/replicaset

we don't have a previous object here to deep equal. We can't perform a deep equal of the pool's VM template spec with the actual VM's spec due to spec defaulting that occurs on the VM and VMI at creation time.

> Also, since this is a label, what will happen if a user modifies the label?

The VM and VMI get updated according to the policies in the VMPool spec.